lib/utils/blockio.js

```diff
@@ -0,0 +1,276 @@
+  bw.writeBytes(checksum);
+  return bw.render();
+};
+
+/**
+ * Cursor to write blocks to the disk
+ * Based on max file size, BlockIO may rollover
+ * this cursor to a new file
+ * @constructor
+ * @param {Number} fd
+ * @param {Number} file
+ * @param {Number} offset
+ */
+
+function FileCursor(fd, file, offset) {
+  if (!(this instanceof FileCursor))
+    return new FileCursor(fd, file, offset);
+
+  this.fd = fd;
+  this.file = file;
+  this.offset = offset;
+}
+
+/**
+ * BlockIO
+ * @alias module:utils.BlockIO
+ * @constructor
+ * @param {Function?} options
+ */
+
+function BlockIO(options) {
+  if (!(this instanceof BlockIO))
+    return new BlockIO(options);
+
+  this.cursor = null;
+  // TODO: limit to 25, implement LRU eviction
+  this.files = {};
+
+  if (options)
+    this.fromOptions(options);
+}
+
+/**
+ * Inject properties from object.
+ * @private
+ * @param {Object} options
+ * @returns {Object}
+ */
+
+BlockIO.prototype.fromOptions = function fromOptions(options) {
+  if (options.network != null)
+    this.network = Network.get(options.network);
+
+  if (options.logger != null) {
+    assert(typeof options.logger === 'object');
+    this.logger = options.logger;
+  }
+
+  if (options.location != null) {
+    assert(typeof options.location === 'string');
+    this.location = options.location;
+  }
+
+  if (options.maxFileSize != null) {
+    assert(util.isU64(options.maxFileSize));
+    this.maxFileSize = options.maxFileSize;
+  }
+
+  return this;
+};
+
+/**
+ * Path to file on disk corresponding to file number
+ * @param {Number} file
+ * @returns {String} - path
+ */
+
+BlockIO.prototype.filepath = function filepath(file) {
+    return path.join(this.location, util.fmt('%d.fdb', file).padStart(13, 0));
+};
+
+/**
+ * Open files
+ * @param {Number?} file (default=0)
+ * @param {Number?} offset (default=0)
+ */
+
+BlockIO.prototype.open = function open(file = 0, offset = 0) {
+  // TODO: load cursor state from db
+  const filepath = this.filepath(file);
+  const fd = fs.openSync(filepath, 'a+');
+  this.cursor = new FileCursor(fd, file, offset);
+};
+
+/**
+ * Close files
+ */
+
+BlockIO.prototype.close = function close() {
+  fs.closeSync(this.cursor.fd);
+};
+
+/**
+ * Ensure working directory exists
+ */
+
+BlockIO.prototype.ensure = function ensure() {
+  if (!fs.existsSync(this.location)) {
+    fs.mkdirpSync(this.location);
+  }
+};
+
+/**
+ * Open corresponding file for read/write
+ * @param {Number} file
+ * @returns {Promise} - Returns fd.
+ */
+
+BlockIO.prototype.openReadWriteFile = async function openReadWriteFile(file) {
+  const filepath = this.filepath(file);
+  let fd = null;
+  try {
+    fd = await fs.open(filepath, 'a+');
+  } catch(e) {
+    throw e;
+  }
+  this.files[file] = fd;
+  return fd;
+};
+
+/**
+ * Retrieve a file by number
+ * @param {Number} file
+ * @returns {Promise} - Returns fd.
+ */
+
+BlockIO.prototype.blockFile = async function blockFile(file) {
+  if (file in this.files) {
```

Comment:
No usage of `in`. Use `if (this.files[file] != null)` if you need to.

Reply:
Done.